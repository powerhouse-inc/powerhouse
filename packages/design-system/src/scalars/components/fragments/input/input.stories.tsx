import type { Meta, StoryObj } from "@storybook/react";
import { Input } from "./input";

const meta = {
  title: "Document Engineering/Simple Components/Input",
  component: Input,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    className: {
      control: "text",
    },
    placeholder: {
      control: "text",
    },
    disabled: {
      control: "boolean",
    },
    value: {
      control: "text",
    },
  },
} satisfies Meta<typeof Input>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    placeholder: "Type something...",
  },
};

export const WithValue: Story = {
  args: {
    value: "Hello World",
  },
};

export const Empty: Story = {
  args: {},
};

export const Placeholder: Story = {
  args: {
    placeholder: "Enter your text here",
  },
};

export const Focused: Story = {
  args: {
    placeholder: "This input is focused",
    autoFocus: true,
  },
  parameters: {
    pseudo: { focus: true },
  },
};

export const Disabled: Story = {
  args: {
    placeholder: "This input is disabled",
    disabled: true,
  },
};

export const WithCustomWidth: Story = {
  args: {
    className: "w-[50px]",
  },
};
