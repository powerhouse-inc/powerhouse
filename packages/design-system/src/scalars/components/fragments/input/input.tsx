import * as React from "react";
import { cn } from "@/scalars/lib/utils";

export type InputProps = React.InputHTMLAttributes<HTMLInputElement>;

const inputBaseStyles = cn(
  // Base styles
  "flex h-10 w-full rounded-md text-sm text-gray-900 dark:text-gray-50",
  // Border & Background
  "dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white",
  // Padding
  "px-3 py-2",
  // Placeholder
  "font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500",
  // Focus styles
  "focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 focus-visible:ring-offset-white",
  "dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900",
  // Disabled state
  "disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500",
);

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  ({ className, type, ...props }, ref) => {
    return (
      <input
        type={type}
        className={cn(inputBaseStyles, className)}
        ref={ref}
        {...props}
      />
    );
  },
);

export { Input };
