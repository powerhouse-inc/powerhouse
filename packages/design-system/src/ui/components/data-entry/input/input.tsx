import * as React from "react";
import { cn } from "../../../../scalars/lib/utils.js";

export type InputProps = React.InputHTMLAttributes<HTMLInputElement>;

const inputBaseStyles = cn(
  // Base styles
  "flex h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50",
  // Border & Background
  "dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white",
  // Padding
  "px-3 py-2",
  // Placeholder
  "font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600",
  // Focus styles
  "focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white",
  "dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50",
  // Disabled state
  "disabled:cursor-not-allowed",
  "disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700",
  "disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300",
);

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  ({ className, type, ...props }, ref) => {
    return (
      <input
        type={type}
        className={cn(inputBaseStyles, className)}
        ref={ref}
        {...props}
      />
    );
  },
);

Input.displayName = "Input";

export { Input };
