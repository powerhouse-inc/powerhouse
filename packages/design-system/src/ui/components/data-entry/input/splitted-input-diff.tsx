import { cn } from "@powerhousedao/design-system/scalars";
import type { WithDifference } from "../../../../scalars/components/types.js";
import { InputDiff } from "./subcomponent/input-diff.js";
import { TextDiff } from "./subcomponent/text-diff.js";

interface SplittedInputDiffProps extends WithDifference<string> {
  value: string;
}

const SplittedInputDiff = ({
  baseValue,
  value,
  viewMode,
  diffMode,
}: SplittedInputDiffProps) => {
  return (
    <InputDiff>
      {viewMode === "mixed" ? (
        <>
          <TextDiff
            baseValue={baseValue}
            value={value}
            viewMode="removal"
            diffMode={diffMode}
            className={cn("flex-1")}
          />
          <div className={cn("ml-3 mr-3 h-[34px] w-px bg-gray-300")} />
          <TextDiff
            baseValue={baseValue}
            value={value}
            viewMode="addition"
            diffMode={diffMode}
            className={cn("flex-1")}
          />
        </>
      ) : (
        <TextDiff
          baseValue={baseValue}
          value={value}
          viewMode={viewMode}
          diffMode={diffMode}
          className={cn("flex-1")}
        />
      )}
    </InputDiff>
  );
};

export { SplittedInputDiff };
