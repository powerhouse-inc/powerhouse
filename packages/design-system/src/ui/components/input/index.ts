export { Input } from "./input.js";
export type { InputProps } from "./input.js";
export { SplittedInputDiff } from "./splitted-input-diff.js";
