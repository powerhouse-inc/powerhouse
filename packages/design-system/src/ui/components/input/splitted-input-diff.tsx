import type { WithDifference } from "@powerhousedao/design-system";
import {
    cn,
    Icon,
    Tooltip,
    TooltipProvider,
} from "@powerhousedao/design-system";
import { useCallback, useState } from "react";
import { useMediaQuery } from "usehooks-ts";
import { InputDiff } from "./subcomponent/input-diff.js";
import { TextDiff } from "./subcomponent/text-diff.js";
interface CopyIconProps {
  value: string;
  hasCopied: boolean;
  setHasCopied: (hasCopied: boolean) => void;
  hasHover: boolean;
}

const CopyIcon = ({
  value,
  hasCopied,
  setHasCopied,
  hasHover,
}: CopyIconProps) => {
  const copy = useCallback(() => {
    navigator.clipboard
      .writeText(value)
      .then(() => {
        setHasCopied(true);
        setTimeout(() => setHasCopied(false), 2000);
      })
      .catch((error: unknown) => {
        console.error("Failed to copy value: ", error);
      });
  }, [value, setHasCopied]);

  return (
    <TooltipProvider>
      <Tooltip content="Copied!" open={hasCopied} triggerAsChild>
        <button
          type="button"
          onClick={copy}
          className={cn(
            "focus-visible:outline-none [&_svg]:pointer-events-none",
            hasHover &&
              "opacity-0 transition-opacity duration-500 group-hover:opacity-100",
          )}
        >
          <Icon name="Copy" size={16} className={cn("text-gray-500")} />
        </button>
      </Tooltip>
    </TooltipProvider>
  );
};

interface SplittedInputDiffProps extends WithDifference<string> {
  value: string;
  showCopyIcon?: boolean;
}

export const SplittedInputDiff = ({
  baseValue,
  value,
  viewMode,
  diffMode,
  showCopyIcon = false,
}: SplittedInputDiffProps) => {
  const [hasCopiedLeft, setHasCopiedLeft] = useState(false);
  const [hasCopiedRight, setHasCopiedRight] = useState(false);
  const hasHover = useMediaQuery("(hover: hover) and (pointer: fine)");

  return (
    <InputDiff className={cn("group")}>
      {viewMode === "mixed" ? (
        <>
          <div
            className={cn(
              "flex flex-1 items-center gap-2 truncate [&>span]:truncate",
            )}
          >
            <TextDiff
              baseValue={baseValue}
              value={value}
              viewMode="removal"
              diffMode={diffMode}
              className={cn("flex-1")}
            />
            {showCopyIcon && baseValue !== undefined && baseValue !== "" && (
              <CopyIcon
                value={baseValue}
                hasCopied={hasCopiedLeft}
                setHasCopied={setHasCopiedLeft}
                hasHover={hasHover}
              />
            )}
          </div>
          <div className={cn("ml-3 mr-3 h-[34px] w-px bg-gray-300")} />
          <div
            className={cn(
              "flex flex-1 items-center gap-2 truncate [&>span]:truncate",
            )}
          >
            <TextDiff
              baseValue={baseValue}
              value={value}
              viewMode="addition"
              diffMode={diffMode}
              className={cn("flex-1")}
            />
            {showCopyIcon && value !== "" && (
              <CopyIcon
                value={value}
                hasCopied={hasCopiedRight}
                setHasCopied={setHasCopiedRight}
                hasHover={hasHover}
              />
            )}
          </div>
        </>
      ) : (
        <div
          className={cn(
            "flex flex-1 items-center gap-2 truncate [&>span]:truncate",
          )}
        >
          <TextDiff
            baseValue={baseValue}
            value={value}
            viewMode={viewMode}
            diffMode={diffMode}
            className={cn("flex-1")}
          />
          {showCopyIcon &&
            ((viewMode === "removal" &&
              baseValue !== undefined &&
              baseValue !== "") ||
              (viewMode === "addition" && value !== "")) && (
              <CopyIcon
                value={viewMode === "removal" ? (baseValue ?? "") : value}
                hasCopied={hasCopiedLeft}
                setHasCopied={setHasCopiedLeft}
                hasHover={hasHover}
              />
            )}
        </div>
      )}
    </InputDiff>
  );
};
