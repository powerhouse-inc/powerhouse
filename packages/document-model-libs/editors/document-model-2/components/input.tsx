import * as React from "react";
import { cn } from "../utils/style.js";

export interface InputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {}

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  ({ className, type, ...props }, ref) => {
    return (
      <input
        type={type}
        className={cn(
          "flex h-9 w-full rounded-md border border-gray-400 bg-transparent px-3 py-1 text-sm transition-colors hover:bg-gray-100  focus-visible:bg-gray-200 focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50",
          className,
        )}
        ref={ref}
        {...props}
      />
    );
  },
);
Input.displayName = "Input";

export { Input };
