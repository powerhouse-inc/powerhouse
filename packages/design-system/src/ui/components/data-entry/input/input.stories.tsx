import type { Meta, StoryObj } from "@storybook/react";
import { Input } from "./input.js";

/**
 * The `Input` component is a basic HTML `input` component that allows users to enter text.
 * It has the standard `input` HTML element attributes and design system styles.
 */
const meta = {
  title: "Document Engineering/Data Entry/Input",
  component: Input,
  parameters: {
    layout: "centered",
    chromatic: {
      disableSnapshot: true,
    },
  },
  tags: ["autodocs"],
  argTypes: {
    className: {
      control: "text",
    },
    placeholder: {
      control: "text",
    },
    disabled: {
      control: "boolean",
    },
    value: {
      control: "text",
    },
  },
} satisfies Meta<typeof Input>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    placeholder: "Type something...",
  },
};

export const WithValue: Story = {
  args: {
    value: "Hello World",
  },
};

export const Empty: Story = {
  args: {},
};

export const Placeholder: Story = {
  args: {
    placeholder: "Enter your text here",
  },
};

export const Focused: Story = {
  args: {
    placeholder: "This input is focused",
    autoFocus: true,
  },
  parameters: {
    pseudo: { focus: true },
  },
};

export const Disabled: Story = {
  args: {
    placeholder: "This input is disabled",
    disabled: true,
  },
};

export const WithCustomWidth: Story = {
  args: {
    className: "w-[50px]",
  },
};
