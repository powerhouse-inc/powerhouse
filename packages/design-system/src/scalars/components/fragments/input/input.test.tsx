import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { Input } from "./input.js";

describe("Input", () => {
  it("should match snapshot", () => {
    const { container } = render(<Input placeholder="Test placeholder" />);
    expect(container).toMatchSnapshot();
  });

  it("should render an input element", () => {
    render(<Input />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("should render with a placeholder", () => {
    render(<Input placeholder="Enter text" />);
    expect(screen.getByPlaceholderText("Enter text")).toBeInTheDocument();
  });

  it("should render with an initial value", () => {
    render(<Input value="Initial text" />);
    expect(screen.getByDisplayValue("Initial text")).toBeInTheDocument();
  });

  it("should be disabled", () => {
    render(<Input disabled />);
    expect(screen.getByRole("textbox")).toBeDisabled();
  });

  it("should accept user input", async () => {
    const user = userEvent.setup();
    render(<Input />);
    const input = screen.getByRole("textbox");

    await user.type(input, "Hello world");
    expect(input).toHaveValue("Hello world");
  });

  it("should call onChange when value changes", async () => {
    const user = userEvent.setup();
    const handleChange = vi.fn();
    render(<Input onChange={handleChange} />);

    await user.type(screen.getByRole("textbox"), "a");
    expect(handleChange).toHaveBeenCalled();
  });

  it("should apply custom className", () => {
    // for testing purposes, we disable the custom class name rule

    render(<Input className="custom-class" />);
    expect(screen.getByRole("textbox")).toHaveClass("custom-class");
  });
});
