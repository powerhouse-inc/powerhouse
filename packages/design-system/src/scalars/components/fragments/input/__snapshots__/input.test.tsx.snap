// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should match snapshot 1`] = `
<div>
  <input
    class="flex h-9 w-full rounded-md text-sm text-gray-900 dark:text-gray-700 border border-gray-300 bg-white dark:border-charcoal-700 dark:bg-charcoal-900 px-3 py-2 font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500"
    placeholder="Test placeholder"
  />
</div>
`;
